tests/material/test_eclblackoilfluidsystemnonstatic.cpp

```diff
@@ -0,0 +1,736 @@
+    // for fluid systems are already tested by the generic test for all fluidsystems.
+
+    using Scalar = typename Opm::MathToolbox<Evaluation>::Scalar;
+    using FluidSystem = Opm::BlackOilFluidSystem<double>;
+
+
+    static constexpr int numPhases = FluidSystem::numPhases;
+
+    static constexpr int gasPhaseIdx = FluidSystem::gasPhaseIdx;
+    static constexpr int oilPhaseIdx = FluidSystem::oilPhaseIdx;
+    static constexpr int waterPhaseIdx = FluidSystem::waterPhaseIdx;
+
+    Opm::Parser parser;
+
+    auto deck = parser.parseString(deckString1);
+    auto python = std::make_shared<Opm::Python>();
+    Opm::EclipseState eclState(deck);
+    Opm::Schedule schedule(deck, eclState, python);
+
+    FluidSystem::initFromState(eclState, schedule);
+
+    auto& fluidSystem = FluidSystem::getNonStaticInstance();
+
+    // NOTE: We want to create the parameter cache from the static version to make sure it works
+    // with the nonstatic version as well.
+    using ParamCache = typename FluidSystem::template ParameterCache<Scalar>;
+    ParamCache paramCache(/*maxOilSat=*/0.5, /*regionIdx=*/1);
+
+    // create a parameter cache
+    BOOST_CHECK_EQUAL(FluidSystem::reservoirTemperature(), fluidSystem.reservoirTemperature());
+    BOOST_CHECK(FluidSystem::enableDissolvedGas());
+    BOOST_CHECK(FluidSystem::enableVaporizedOil());
+    BOOST_CHECK_EQUAL(FluidSystem::numRegions(), 2);
+    BOOST_CHECK_EQUAL(FluidSystem::numActivePhases(), 3);
+
+    for (unsigned phaseIdx = 0; phaseIdx < numPhases; ++phaseIdx) {
+        BOOST_CHECK(FluidSystem::phaseIsActive(phaseIdx));
+    }
+
+    BOOST_CHECK_EQUAL(FluidSystem::solventComponentIndex(oilPhaseIdx), fluidSystem.solventComponentIndex(oilPhaseIdx));
+    BOOST_CHECK_EQUAL(FluidSystem::solventComponentIndex(gasPhaseIdx), fluidSystem.solventComponentIndex(gasPhaseIdx));
+    BOOST_CHECK_EQUAL(FluidSystem::solventComponentIndex(waterPhaseIdx),
+                      fluidSystem.solventComponentIndex(waterPhaseIdx));
+
+    BOOST_CHECK_EQUAL(FluidSystem::soluteComponentIndex(oilPhaseIdx), fluidSystem.soluteComponentIndex(oilPhaseIdx));
+    BOOST_CHECK_EQUAL(FluidSystem::soluteComponentIndex(gasPhaseIdx), fluidSystem.soluteComponentIndex(gasPhaseIdx));
+
+    BOOST_CHECK_EQUAL(FluidSystem::referenceDensity(oilPhaseIdx, /*regionIdx=*/1),
+                      fluidSystem.referenceDensity(oilPhaseIdx, /*regionIdx=*/1));
+    BOOST_CHECK_EQUAL(FluidSystem::referenceDensity(gasPhaseIdx, /*regionIdx=*/1),
+                      fluidSystem.referenceDensity(gasPhaseIdx, /*regionIdx=*/1));
+    BOOST_CHECK_EQUAL(FluidSystem::referenceDensity(waterPhaseIdx, /*regionIdx=*/1),
+                      fluidSystem.referenceDensity(waterPhaseIdx, /*regionIdx=*/1));
+
+    Opm::BlackOilFluidState<Scalar, FluidSystem> fluidState {};
+    Opm::Valgrind::SetUndefined(fluidState);
+
+    static const Scalar eps = std::sqrt(std::numeric_limits<Scalar>::epsilon());
+    unsigned regionIdx = paramCache.regionIndex();
```

Comment:
As far as I can tell there are two pvt regions, suggest to loop over both instead of just doing region 1.